### PR TITLE
estimate_cox.R - wrapper for cox model

### PR DESCRIPTION
**Pull request by timtreis**
Hey Alexandra, I moved this file into a pull request until it is fully functional and tested :) Then we can merge it again.

I already fixed a few minor things, do you want to take over again?

### CODE REVIEW

File: R/estimate_cox.R [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+#'
+#' ## Univariate Cox regression model using the covariate: `TRTP`
+#' # Stratified Kaplan-Meier analysis by `TRTP`
+#' visR::estimate_cox(data = adtte, main = "TRTP")
+#'
+#' ## Multivariate Cox regression model using the covariates: `TRTP` and `SEX`
+#' visR::estimate_cox(data = adtte, main = c("TRTP", "SEX"))
+#'
+#' ## Stratified Cox model (strata with one level)
+#' visR::estimate_cox(data = adtte, main = c("AGE", "survival::strata(PARAMCD)"))
```
Comment (SHAESEN2) [inline R/estimate_cox.R:36]: i would add different argument for strata and for interactions

File: R/estimate_cox.R [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+#'                CNSR = dplyr::if_else(status == 1, 0, 1)
+#'  )
+#'
+#' visR::estimate_cox(data = veteran_adam, main = c("trt", "diagtime"))
+
+
+
+estimate_cox <- function(
+  data = NULL
+  ,main = NULL
```
Comment (SHAESEN2) [inline R/estimate_cox.R:61]: if you want to put all main, modifiers and interactions into one statment, then opt for consistency with estimate_KM and  call it strata

File: R/estimate_cox.R [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    stop(paste0("data can be of class `data.frame` or `tibble` or `data.table`."))
+  }
+
+  data <- as.data.frame(data)
+
+  # Validate columns --------------------------------------------------------
+
+  reqcols <- c(main, CNSR, AVAL)
+
+  # if (! all(reqcols %in% colnames(data))){
```
Comment (SHAESEN2) [inline R/estimate_cox.R:99]: Why is this removed?

File: R/estimate_cox.R [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+  }
+
+  if (!is.numeric(data[[CNSR]])) {
+    stop("Censor variable (CNSR) is not numeric.")
+  }
+
+
+  # Remove NA from the analysis --------------------------------------------
+
+  data <- as.data.frame(data) %>%
```
Comment (SHAESEN2) [inline R/estimate_cox.R:114]: main should also be here

File: R/estimate_cox.R [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    survfit_object[["PARAMCD"]] <- paste(unique(data[["PARAMCD"]]), collapse = ", ")
+  }
+
+
+  # Return ------------------------------------------------------------------
+
+
+  summary_cox <- summary(survfit_object, conf.int = conf.int)
+
+ # res1 <- c(
```
Comment (SHAESEN2) [inline R/estimate_cox.R:175]: The next section could go in a tidyme method

**Comment by AlexandraP-21** [comment]
@timtreis thanks for noticing, this is because PARAMCD has only 1 category. I will change it to RACE which has more.

File: R/estimate_cox.R (hunk 2 above)
Comment (AlexandraP-21) [reply R/estimate_cox.R:61]: We cannot call it strata because strata we can have stratified parameters which will be denoted with 'strata' in the model. So we cannot have strata for the stratified parameters and strata for the model as well. But other ideas instead of main are more than welcome! :)

File: R/estimate_cox.R (hunk 3 above)
Comment (AlexandraP-21) [reply R/estimate_cox.R:99]: Because 'main' might have 'strata(RACE)' so that means that the if statement will start looking for 'strata(RACE)' and not just 'RACE'. I was trying to find a way to just get the parameters but I could not figure something out. Again any ideas more than welcome!

File: R/estimate_cox.R (hunk 2 above)
Comment (SHAESEN2) [reply R/estimate_cox.R:61]: We can have both. eg model(data=data) Can you give an example of why it would not work?

File: R/estimate_cox.R (hunk 3 above)
Comment (SHAESEN2) [reply R/estimate_cox.R:99]: Can you give a better example?

File: R/estimate_cox.R [hunk 6]
```diff
@@ -0,0 +1,181 @@
+#' @title Wrapper for Cox Proportional Hazard analysis
+#'
+#' @description This function is a wrapper around \code{survival::coxph} to fit a Cox Proportional Hazards Regression model, assuming right-censored data.
+#'    The function expects that the data has been filtered on the parameter (PARAM/PARAMCD) of interest.
+#'    The result is an object of class \code{summary.coxph} which can be used in downstream functions and methods that rely on the \code{summary.coxph} class.
+#'
```
Comment (SHAESEN2) [inline R/estimate_cox.R:6]: #'    By default:
#'    \itemize{
#'      \item{A two-sided pointwise 0.95 confidence interval is estimated using a log transformation (conf.type = "log").}
#'      \item{The Efron approximation is used for tie handling.}
#'    }

@AlexandraP-21 Should we enforce some other SAS defaults? What are some of the other differences with SAS? Should we document them in this description as well? I believe it might be important for Pharma to understand those without investigating it themselves as we did. We have a presentation about those differences ;)

File: R/estimate_cox.R [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+#' @param data The name of the dataset for the Cox model is based on the Analysis Data Model (ADaM) principles. The dataset is expected to have
+#'    one record per subject per analysis parameter. Rows in which the analysis variable (AVAL) or the sensor variable (CNSR) contain NA, are removed during analysis.
+#' @param equation Character vector, representing the covariates used at the Cox model. When NULL, the baseline (null) model is fit.
```
Comment (SHAESEN2) [inline R/estimate_cox.R:9]: #' @param equation Character vector, representing the formula to be estimated with the Cox model. When NULL, the baseline (null) model is fit.

File: R/estimate_cox.R [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+  ,AVAL = "AVAL"
+  ,conf.int = 0.95
+  ,...
+){
+
+  # Capture input to validate user input for data argument -----------------
+
+  ## Get actual data name as symbol
+  ### Magrittre pipe returns "." which inactivates recalls to survfit in downstream functions
+  ### map passes .x as cox_ccall$data
```
Comment (SHAESEN2) [inline R/estimate_cox.R:72]: ### map passes .x as cox_call$data
  
  There was a typo

File: R/estimate_cox.R [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+  if (!(inherits(data, "data.frame") | inherits(data, "tibble") | inherits(data, "data.table"))) {
+    stop(paste0("data can be of class `data.frame` or `tibble` or `data.table`."))
+  }
+
+  data <- as.data.frame(data)
+
+  # Validate columns --------------------------------------------------------
+
+  reqcols <- c(equation, CNSR, AVAL)
+  ss <- c("strata", "tt", "frailty", "ridge", "pspline", "cluster", "I", "factor")
```
Comment (SHAESEN2) [inline R/estimate_cox.R:98]: I dont like following code because the function will evolve and some arguments in coxph will get depricated.

I would replace this part with something more defensive or don't overengineer it:
available <- names(as.list(args(survival::coxph)))

requested <- names(as.list(args(survival::survfit)))

intersect(requested, available)

File: R/estimate_cox.R [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+  data <- as.data.frame(data) %>%
+    tidyr::drop_na(AVAL, CNSR)
+
+
+  # Ensure the presence of at least one strata -----------------------------
+
+  if (is.null(equation)) {
+    equation <- "1"
+  } else {
+    equation <- paste(equation, collapse = " + ")
```
Comment (SHAESEN2) [inline R/estimate_cox.R:132]: For the equation, I would let the user fill it out completely. I found it very confusing to understand the examples untill I saw the function collapses it. 
For me it does not make sense for the cox. If I have to fill out an equation, I dont want to think about how it is interpreted by the downstream function. Therefore, I would remove the following code:

  } else {
    equation <- paste(equation, collapse = " + ")
  }

File: R/estimate_cox.R [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+  }
+
+
+  ## Reverse censoring: see ADaM guidelines versus R survival KM analysis
+  library(survival)
```
Comment (SHAESEN2) [inline R/estimate_cox.R:137]: remove line of code

File: R/estimate_cox.R [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+    names(dots)
+    for (i in seq_along(dots)) {
+      survfit_object$call[[names(dots)[i]]] <- unlist(dots[i], use.names = FALSE)
+    }
+  }
+
+
+  # Add additional metadata -------------------------------------------------
+
+  if ("PARAM" %in% colnames(data) && length(setdiff(c("PARAMCD", "PARAM"), equation)) == 2) {
```
Comment (SHAESEN2) [inline R/estimate_cox.R:161]: revise based on other comments

File: R/estimate_cox.R [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+    # we expect only one unique value => catch mistakes
+    survfit_object[["PARAM"]] <- paste(unique(data[["PARAM"]]), collapse = ", ")
+  }
+
+  if ("PARAMCD" %in% colnames(data) && length(setdiff(c("PARAMCD", "PARAM"), equation)) == 2) {
+    # we expect only one unique value => catch mistakes
+    survfit_object[["PARAMCD"]] <- paste(unique(data[["PARAMCD"]]), collapse = ", ")
+  }
+
```
Comment (SHAESEN2) [inline R/estimate_cox.R:170]: Dont we need to add 'Overall'/Null Model as strata when we dont fill in equation? When we plot the null model this needs to be clear that this is the null model as described in your documentation

File: R/estimate_cox.R [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+
+  # Return ------------------------------------------------------------------
+
+
+  summary_cox <- summary(survfit_object, conf.int = conf.int)
```
Comment (SHAESEN2) [inline R/estimate_cox.R:175]: We are creating a wrapper around coxph so you need to return the same object as coxph does, not a summary thereoff. We are allowed to extend the coxph object so if you need extra info from the summary, then add it to the coxph object that you return and document it as such.

**Review by SHAESEN2 — CHANGES_REQUESTED** [review]

File: R/estimate_cox.R (hunk 7 above)
Comment (AlexandraP-21) [reply R/estimate_cox.R:9]: Done

File: R/estimate_cox.R (hunk 10 above)
Comment (AlexandraP-21) [reply R/estimate_cox.R:132]: Done

**Comment by SHAESEN2** [comment]
add in reexports.R file export from survival STRATA blah

**Comment by SHAESEN2** [comment]
see #362
